File: 프로그래머스/N으로 표현/차애리.cpp
#include <iostream>
#include <unordered_set>
#include <vector>
#include <cmath>
#include <string>

using namespace std;


int solution(int N, int number){
    int answer = -1;
    int MIN = 8;
    int base = 0;
    unordered_set<int> numbers[MIN];
    for(int i = 0; i < MIN; i++){
        base = 10 * base + 1;
        numbers[i].insert(base*N);
    }

    for(int i = 0; i< MIN; i++){
        for(int j = 0; j <i ; j++){
            for (auto op1 : numbers[j]){
                for(auto op2 : numbers[i-j-1]){
                    numbers[i].insert(op1+op2);
                    numbers[i].insert(op1-op2);
                    numbers[i].insert(op1*op2);
                    if(op2!=0){
                        numbers[i].insert(op1/op2);
                    }

                }
            }
        }
        if(numbers[i].count(number) >0){
            answer = i + 1;
            break;
        }
    }
    return answer;

}

File: 프로그래머스/N으로 표현/차애리.py
def solution(N, number):
    answer = -1
    # i개를 이용한 조합 =>
    #i* str(N) +  N*(i-1) + (i-1)로 만든 사칙연산 조합, (i-1)로 만든 사칙연산 조합 + N*(i-1)
    # N을 i만큼 사용 +    a   +              b       또는        a               +      b
    
    DP = [] #각 i번째 계산된 값들의 집합이 중복없이 저장됨
    for i in range(1,9):
        hubo = set() # 각 i번째 조합들이 저장됨
        hubo.add(int(str(N)*i)) # N을 i만큼 사용
        
        for j in range(0,i-1):
            for x in DP[j]: #(i-1)로 만든 조합
                for y in DP[-j-1]:#(i-1)로 만든 조합(뒤에서)
                    hubo.add(x+y)
                    hubo.add(x-y)
                    hubo.add(x*y)
                    
                    if(y !=0):
                        hubo.add(x//y)      
        
        if ( number in hubo):
            answer = i
            break
        DP.append(hubo) # 동적계획법 Memorization 사용
        
    
    
    return answer
